com.hazelcast.jet.grpc.GrpcServiceTest failed due to OutOfMemoryError

_master_ (commit edb1e3c05c628881f9509a5277ed786e70b948c2)

Failed on Sonar Nightly build: http://jenkins.hazelcast.com/job/jet-oss-master-sonar-nightly/458/testReport/junit/com.hazelcast.jet.grpc/GrpcServiceTest/com_hazelcast_jet_grpc_GrpcServiceTest/

Stacktrace:
```
org.junit.runners.model.TestTimedOutException: test timed out after 900 seconds
	at sun.management.ThreadImpl.getThreadUserTime(ThreadImpl.java:293)
	at sun.management.ThreadImpl.getThreadUserTime(ThreadImpl.java:285)
	at com.hazelcast.test.AbstractHazelcastClassRunner.generateThreadDump(AbstractHazelcastClassRunner.java:351)
	at com.hazelcast.test.AbstractHazelcastClassRunner.access$000(AbstractHazelcastClassRunner.java:64)
	at com.hazelcast.test.AbstractHazelcastClassRunner$ThreadDumpAwareRunAfters.evaluate(AbstractHazelcastClassRunner.java:394)
	at com.hazelcast.test.DumpBuildInfoOnFailureRule$1.evaluate(DumpBuildInfoOnFailureRule.java:37)
	at com.hazelcast.test.jitter.JitterRule$1.evaluate(JitterRule.java:104)
	at com.hazelcast.test.metrics.MetricsRule$1.evaluate(MetricsRule.java:62)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at com.hazelcast.test.HazelcastSerialClassRunner.runChild(HazelcastSerialClassRunner.java:50)
	at com.hazelcast.test.HazelcastSerialClassRunner.runChild(HazelcastSerialClassRunner.java:29)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at com.hazelcast.test.AbstractHazelcastClassRunner$1.evaluate(AbstractHazelcastClassRunner.java:301)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

Standard output:
```
2020-04-12 13:16:15,846 [DEBUG] [hz.amazing_keller.cached.thread-43] [c.h.j.i.MasterJobContext]: [127.0.0.1]:5701 [jet] [4.1-SNAPSHOT] Not scaling up job '042f-e0f2-edc0-0001', execution 042f-e0f2-edc1-0001: not running or already running on all members
...
2020-04-12 13:18:18,680 [DEBUG] [hz.amazing_keller.cached.thread-24] [c.h.j.i.JobCoordinationService]: [127.0.0.1]:5701 [jet] [4.1-SNAPSHOT] Not starting jobs because master is not in safe state.
...
2020-04-12 13:19:23,469 [ERROR] [hz.amazing_keller.partition-operation.thread-60] [c.h.s.i.o.i.OperationExecutorImpl]: [127.0.0.1]:5701 [jet] [4.1-SNAPSHOT] Failed to process: com.hazelcast.spi.impl.operationexecutor.impl.TaskBatch@348efd57 on: hz.amazing_keller.partition-operation.thread-60
java.lang.OutOfMemoryError: Java heap space
	at java.nio.HeapByteBuffer.<init>(HeapByteBuffer.java:57) ~[?:1.8.0_241]
	at java.nio.ByteBuffer.allocate(ByteBuffer.java:335) ~[?:1.8.0_241]
	at com.hazelcast.test.mocknetwork.MockConnection.readFromPacket(MockConnection.java:112) ~[hazelcast-4.0.1-SNAPSHOT-tests.jar:4.0.1-SNAPSHOT]
	at com.hazelcast.test.mocknetwork.MockConnection.write(MockConnection.java:99) ~[hazelcast-4.0.1-SNAPSHOT-tests.jar:4.0.1-SNAPSHOT]
	at com.hazelcast.test.mocknetwork.MockNetworkingService$MockEndpointManager.transmit(MockNetworkingService.java:246) ~[hazelcast-4.0.1-SNAPSHOT-tests.jar:4.0.1-SNAPSHOT]
	at com.hazelcast.test.mocknetwork.MockNetworkingService$MockEndpointManager.transmit(MockNetworkingService.java:90) ~[hazelcast-4.0.1-SNAPSHOT-tests.jar:4.0.1-SNAPSHOT]
	at com.hazelcast.internal.nio.tcp.FirewallingNetworkingService$FirewallingEndpointManager.transmit(FirewallingNetworkingService.java:291) ~[hazelcast-4.0.1-SNAPSHOT-tests.jar:4.0.1-SNAPSHOT]
	at com.hazelcast.spi.impl.operationservice.impl.OutboundOperationHandler.send(OutboundOperationHandler.java:64) ~[hazelcast-4.0.1-SNAPSHOT.jar:4.0.1-SNAPSHOT]
	at com.hazelcast.spi.impl.operationservice.impl.OutboundOperationHandler.send(OutboundOperationHandler.java:52) ~[hazelcast-4.0.1-SNAPSHOT.jar:4.0.1-SNAPSHOT]
	at com.hazelcast.spi.impl.operationservice.impl.OperationServiceImpl.send(OperationServiceImpl.java:444) ~[hazelcast-4.0.1-SNAPSHOT.jar:4.0.1-SNAPSHOT]
	at com.hazelcast.internal.partition.impl.AbstractPartitionPrimaryReplicaAntiEntropyTask.invokePartitionBackupReplicaAntiEntropyOp(AbstractPartitionPrimaryReplicaAntiEntropyTask.java:115) ~[hazelcast-4.0.1-SNAPSHOT.jar:4.0.1-SNAPSHOT]
	at com.hazelcast.internal.partition.impl.PartitionPrimaryReplicaAntiEntropyTask.run(PartitionPrimaryReplicaAntiEntropyTask.java:46) ~[hazelcast-4.0.1-SNAPSHOT.jar:4.0.1-SNAPSHOT]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:177) ~[hazelcast-4.0.1-SNAPSHOT.jar:4.0.1-SNAPSHOT]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:192) ~[hazelcast-4.0.1-SNAPSHOT.jar:4.0.1-SNAPSHOT]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:144) [hazelcast-4.0.1-SNAPSHOT.jar:4.0.1-SNAPSHOT]
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:123) [hazelcast-4.0.1-SNAPSHOT.jar:4.0.1-SNAPSHOT]
```

Standard Error:
```
Exception in thread "grpc-default-executor-3699" Exception in thread "surefire-forkedjvm-command-thread" Exception in thread "grpc-default-executor-974" Exception in thread "grpc-default-executor-3594" Exception in thread "grpc-default-executor-958" java.lang.OutOfMemoryError: Java heap space
java.lang.OutOfMemoryError: Java heap spaceTHREAD DUMP FOR TEST FAILURE: "test timed out after 300000 milliseconds" at "when_unary_distributed"


Exception in thread "grpc-default-executor-980" Exception in thread "grpc-default-executor-976" java.lang.OutOfMemoryError: Java heap space
Exception in thread "grpc-default-executor-6186" java.lang.OutOfMemoryError: Java heap space
Exception in thread "grpc-default-executor-4914" Exception in thread "grpc-default-executor-5260" java.lang.OutOfMemoryError: Java heap space
java.lang.OutOfMemoryError: Java heap space
java.lang.OutOfMemoryError: Java heap space
java.lang.OutOfMemoryError: Java heap space
java.lang.OutOfMemoryError: Java heap space
java.lang.OutOfMemoryError: Java heap space
```

Content of generated dumpstream file:
```
# Created at 2020-04-12T13:17:12.409
*** java.lang.instrument ASSERTION FAILED ***: "!errorOutstanding" with message can't create byte arrau at JPLISAgent.c line: 813

# Created at 2020-04-12T13:24:40.306


# Created at 2020-04-12T13:24:40.306
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "grpc-default-executor-4827"

# Created at 2020-04-12T13:24:42.494


# Created at 2020-04-12T13:24:42.494
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "hz.amazing_keller.jet.cooperative.thread-4"

# Created at 2020-04-12T13:24:43.402


# Created at 2020-04-12T13:24:43.403
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "grpc-default-executor-4218"

# Created at 2020-04-12T13:24:44.317


# Created at 2020-04-12T13:24:44.317
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "grpc-default-executor-1109"

# Created at 2020-04-12T13:24:45.190


# Created at 2020-04-12T13:24:45.190
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "grpc-default-executor-778"

# Created at 2020-04-12T13:26:43.299


# Created at 2020-04-12T13:26:43.299
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "grpc-default-executor-4089"

# Created at 2020-04-12T13:26:43.783


# Created at 2020-04-12T13:26:43.783
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "grpc-default-executor-2766"

# Created at 2020-04-12T13:26:44.275


# Created at 2020-04-12T13:26:44.275
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "grpc-default-executor-2891"

# Created at 2020-04-12T13:26:46.735


# Created at 2020-04-12T13:26:46.735
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "grpc-default-executor-3597"

# Created at 2020-04-12T13:26:48.978


# Created at 2020-04-12T13:26:48.979
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "grpc-default-executor-2823"

# Created at 2020-04-12T13:26:50.186


# Created at 2020-04-12T13:26:50.186
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "grpc-default-executor-4083"

# Created at 2020-04-12T13:26:54.197


# Created at 2020-04-12T13:26:54.197
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "grpc-default-executor-3596"

# Created at 2020-04-12T13:26:55.060


# Created at 2020-04-12T13:26:55.060
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "grpc-default-executor-4087"

# Created at 2020-04-12T13:26:56.168


# Created at 2020-04-12T13:26:56.168
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "grpc-default-executor-3590"

# Created at 2020-04-12T13:26:59.901


# Created at 2020-04-12T13:26:59.901
Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "grpc-default-executor-1149"

```

--- a/examples/grpc/src/main/java/com/hazelcast/jet/examples/grpc/GRPCEnrichment.java
+++ b/examples/grpc/src/main/java/com/hazelcast/jet/examples/grpc/GRPCEnrichment.java
@@ -37,6 +37,7 @@ import java.io.InputStreamReader;
 import java.util.Map;
 import java.util.Map.Entry;
 import java.util.concurrent.CancellationException;
+import java.util.concurrent.Executors;
 import java.util.stream.Stream;
 
 import static com.hazelcast.function.Functions.entryValue;
@@ -86,6 +87,7 @@ public final class GRPCEnrichment {
                 .collect(toMap(Entry::getKey, e -> new Broker(e.getKey(), e.getValue())));
 
         ServerBuilder.forPort(PORT)
+                     .executor(Executors.newFixedThreadPool(4))
                      .addService(new ProductServiceImpl(productMap))
                      .addService(new BrokerServiceImpl(brokerMap))
                      .build()

--- a/extensions/grpc/src/test/java/com/hazelcast/jet/grpc/GrpcServiceTest.java
+++ b/extensions/grpc/src/test/java/com/hazelcast/jet/grpc/GrpcServiceTest.java
@@ -41,6 +41,7 @@ import org.junit.Test;
 import java.io.IOException;
 import java.util.Arrays;
 import java.util.List;
+import java.util.concurrent.Executors;
 import java.util.concurrent.TimeUnit;
 import java.util.stream.IntStream;
 
@@ -236,7 +237,10 @@ public class GrpcServiceTest extends SimpleTestInClusterSupport {
     }
 
     private static Server createServer(BindableService service) throws IOException {
-        Server server = ServerBuilder.forPort(0).addService(service).build();
+        Server server = ServerBuilder.forPort(0)
+                                     .executor(Executors.newFixedThreadPool(4))
+                                     .addService(service)
+                                     .build();
         server.start();
         return server;
     }

